### PR TITLE
Replacing MockReateLimiter with mocking object to improve test design

### LINKED ISSUE #2253: Refactor MockRateLimiter to improve test design
### Description

I noticed that there is a test class [MockRateLimiter](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/test/java/org/apache/accumulo/core/file/streams/MockRateLimiter.java#L25) implements a production interface [RateLimiter](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/main/java/org/apache/accumulo/core/util/ratelimit/RateLimiter.java#L21) to assist testing production function [RateLimitedOutputStream.write(byte[], int, int)](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/main/java/org/apache/accumulo/core/file/streams/RateLimitedOutputStream.java#L47) and [RateLimitedInputStream.read(byte[], int, int)](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/main/java/org/apache/accumulo/core/file/streams/RateLimitedInputStream.java#L51). This might not be the best priactice in unit testing and can be improved by leveraging mocking frameworks. 

### Current Implementation
- `MockRateLimiter` use AtomicLong variable to keep tracking excution times of [RateLimiter:acquire(long)](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/main/java/org/apache/accumulo/core/util/ratelimit/RateLimiter.java#L28).
- Use assertion statements in unit tests to verify the behavior of the `MockRateLimiter` to make sure `read` and `write` function excuted as desired.

### Proposed Implementation
- Replace `MockRateLimiter` with a mocking object created by Mockito.
- Use method stub to control the behavior of the mocking object.
- Extrac variables in test case to make test condition more explict.

### Motivation

- Decoupling test class `MockRateLimiter` from production class `RateLimiter`.
- Making test condition more clear by verifying mocking obejct's behavior with a local variable.
- Making test logic more clear by using method stub instead of method overriding.

### PR DESCRIPTION
**Pull request by wx930910**
Fixes #2253

#### Refactor test class [MockRateLimiter](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/test/java/org/apache/accumulo/core/file/streams/MockRateLimiter.java#L25) and test file [RateLimitedInputStreamTest.java](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java#L33), [RateLimitedOutputStreamTest.java](https://github.com/apache/accumulo/blob/1dc72fce2c781dee597c8c11876a3bc6c321c199/core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedOutputStreamTest.java#L35)

<hr>

##### Motivation

- Decoupling test class `MockRateLimiter` from production class `RateLimiter`.
- Making test condition more clear by verifying mocking obejct's behavior with a local variable.
- Making test logic more clear by using method stub instead of method overriding.

<hr>

##### Key changed/added classes in this PR
 * Created mocking object to replace test subclass `MockRateLimiter`, decoupled test from production code.
 * Extract `PermitsAcquired` as a local variable to improve test logic and make test condition more explict.

<hr>

### CODE REVIEW

**Comment by milleruntime** [comment]
@wx930910 thanks for the contribution. If you run the build using "mvn clean verify -DskipTests" it will take care of a bunch of code quality assurance checks automatically.

**Comment by DomGarguilo** [comment]
I don't think accumulo has mockito as a depenency yet. It does have EasyMock though. Does it seem like what you are trying to do with mockito can be done with easymock?

**Comment by wx930910** [comment]
> @wx930910 thanks for the contribution. If you run the build using "mvn clean verify -DskipTests" it will take care of a bunch of code quality assurance checks automatically.

Thanks! Will do that and revise the code accordingly!

**Comment by wx930910** [comment]
> I don't think accumulo has mockito as a depenency yet. It does have EasyMock though. Does it seem like what you are trying to do with mockito can be done with easymock?

No, we do not have Mockito dependency in Accumulo. I think EasyMock can do similar logic and I will migrate the code change to EasyMock.

**Comment by wx930910** [comment]
@DomGarguilo I revised the code and use similar logic to create mock object by EasyMock, please let me know if you have any other suggestions!

**Comment by EdColeman** [comment]
The build failed because of formatting issues - as said earlier, if you build locally it will automatically reformat the code.  Most of the QA checks and your tests can be run using:

```mvn clean verify javadoc:jar -Psec-bugs -DskipITs=true -Dtest=RateLimitedInputStreamTest,RateLimitedOutputStreamTest```

Some of these checks will add to the build time so you might want to run the full set when you are ready to commit.  The original suggestion:

```mvn clean verify -DskipTests```

Should correct your current issue.  I was able to get the full set to pass from your current branch.

**Comment by wx930910** [comment]
@EdColeman Thank you for your help! I reformat the code and now it passed the QA check.

File: core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java [hunk 1]
```diff
@@ -56,22 +69,22 @@ public int read() {
     @Override
     public void seek(long pos) {
       throw new UnsupportedOperationException("Not supported yet."); // To change body of generated
```
Comment (EdColeman) [inline core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java:71]: The comment seems to be from you IDE and not relevant to the code.  Please update or remove,

File: core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java [hunk 2]
```diff
@@ -56,22 +69,22 @@ public int read() {
...
-                                                                     // methods, choose Tools |
-                                                                     // Templates.
+      // methods, choose Tools |
+      // Templates.
     }
 
     @Override
     public long getPos() {
       throw new UnsupportedOperationException("Not supported yet."); // To change body of generated
```
Comment (EdColeman) [inline core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java:78]: The comment seems to be from you IDE and not relevant to the code.  Please update or remove,

File: core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java [hunk 3]
```diff
@@ -56,22 +69,22 @@ public int read() {
...
-                                                                     // methods, choose Tools |
-                                                                     // Templates.
+      // methods, choose Tools |
+      // Templates.
     }
 
     @Override
     public boolean seekToNewSource(long targetPos) {
       throw new UnsupportedOperationException("Not supported yet."); // To change body of generated
```
Comment (EdColeman) [inline core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java:85]: The comment seems to be from you IDE and not relevant to the code.  Please update or remove,

Comment (wx930910) [reply core/src/test/java/org/apache/accumulo/core/file/streams/RateLimitedInputStreamTest.java]: @EdColeman Thanks! The comments existed in code repository before this change. Now I removed them.

**Review by dlmarion — APPROVED** [review]

**Review by DomGarguilo — APPROVED** [review]

**Review by milleruntime — APPROVED** [review]
Thanks for the contribution @wx930910 and welcome to the Apache Accumulo community. If you would like to be listed as a contributor on Accumulo's people page please make a website PR for people.md. https://github.com/apache/accumulo-website/edit/main/pages/people.md

If you intend to make more contributions, please consider subscribing to the dev list and introducing yourself. Also, you can use the dev list or slack to reach out if you have questions about anything. https://accumulo.apache.org/contact-us/

**Comment by wx930910** [comment]
@milleruntime Thank you so much for helping me out! I just updated the people page and would love to make more contributions!

**Pull request merged**